### PR TITLE
GH-2209 Reproduce Changeset concurrent modification exception

### PR DESCRIPTION
**Pull request by hmottestad**
Briefly describe the changes proposed in this PR:

Reproduces: GH-2209

This reproduces the issue that we are seeing in Jenkins every once in a while.

---- 
PR Author Checklist: 

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [ ] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

### CODE REVIEW

**Comment by hmottestad** [comment]
This even happens without parallel validation (eg. sharing a single connection between multiple threads). So it's truly an issue separate from the ShaclSail, just hasn't been shown before because we don't have any other tests of this intensity.

**Comment by hmottestad** [comment]
```
2020-05-10T21:37:15.4773575Z java.util.ConcurrentModificationException
2020-05-10T21:37:15.4774490Z 	at java.base/java.util.LinkedHashMap$LinkedHashIterator.nextNode(LinkedHashMap.java:719)
2020-05-10T21:37:15.4775890Z 	at java.base/java.util.LinkedHashMap$LinkedKeyIterator.next(LinkedHashMap.java:741)
2020-05-10T21:37:15.4777402Z 	at org.eclipse.rdf4j.model.util.PatternIterator.findNextElement(PatternIterator.java:68)
2020-05-10T21:37:15.4778809Z 	at org.eclipse.rdf4j.model.util.PatternIterator.hasNext(PatternIterator.java:46)
2020-05-10T21:37:15.4780565Z 	at org.eclipse.rdf4j.model.impl.LinkedHashModel$ModelIterator.hasNext(LinkedHashModel.java:312)
2020-05-10T21:37:15.4781978Z 	at org.eclipse.rdf4j.model.impl.LinkedHashModel.contains(LinkedHashModel.java:205)
2020-05-10T21:37:15.4783645Z 	at org.eclipse.rdf4j.sail.nativerdf.MemoryOverflowModel.contains(MemoryOverflowModel.java:117)
2020-05-10T21:37:15.4784947Z 	at org.eclipse.rdf4j.model.impl.DynamicModel.contains(DynamicModel.java:101)
2020-05-10T21:37:15.4786040Z 	at org.eclipse.rdf4j.sail.base.Changeset.prepare(Changeset.java:116)
2020-05-10T21:37:15.4787382Z 	at org.eclipse.rdf4j.sail.base.SailSourceBranch$1.prepare(SailSourceBranch.java:169)
2020-05-10T21:37:15.4788732Z 	at org.eclipse.rdf4j.sail.base.SailSourceBranch.prepare(SailSourceBranch.java:246)
2020-05-10T21:37:15.4789856Z 	at org.eclipse.rdf4j.sail.base.UnionSailSource.prepare(UnionSailSource.java:62)
2020-05-10T21:37:15.4827778Z 	at org.eclipse.rdf4j.sail.base.SailSourceConnection.prepareInternal(SailSourceConnection.java:457)
2020-05-10T21:37:15.4828681Z 	at org.eclipse.rdf4j.sail.helpers.AbstractSailConnection.prepare(AbstractSailConnection.java:365)
2020-05-10T21:37:15.4830186Z 	at org.eclipse.rdf4j.sail.shacl.ShaclSailConnection.prepare(ShaclSailConnection.java:647)
2020-05-10T21:37:15.4832071Z 	at org.eclipse.rdf4j.repository.sail.SailRepositoryConnection.commit(SailRepositoryConnection.java:161)
2020-05-10T21:37:15.4834032Z 	at org.eclipse.rdf4j.sail.shacl.MultithreadedTest.lambda$runValidationFailuresTest$3(MultithreadedTest.java:497)
2020-05-10T21:37:15.4835681Z 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
2020-05-10T21:37:15.4837020Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2020-05-10T21:37:15.4838672Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
2020-05-10T21:37:15.4840282Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
2020-05-10T21:37:15.4840799Z 	at java.base/java.lang.Thread.run(Thread.java:834)

```

**Comment by hmottestad** [comment]
@jeenbroekstra finally tracked down the bug causing many of the unstable builds in Jenkins. You can try it out for yourself in the multi threaded native store test in the SHACL package.

**Comment by hmottestad** [comment]
@jeenbroekstra this is the code that reproduces the multithreaded model access that is present for serializable transaction on the native store.

File: core/model/src/main/java/org/eclipse/rdf4j/model/impl/LinkedHashModel.java [hunk 1]
```diff
@@ -200,7 +200,9 @@ public Iterator iterator() {
 
 	@Override
 	public boolean contains(Resource subj, IRI pred, Value obj, Resource... contexts) {
-		return matchPattern(subj, pred, obj, contexts).hasNext();
+		ModelIterator modelIterator = matchPattern(subj, pred, obj, contexts);
+		Thread.yield();
+		return modelIterator.hasNext();
```
Comment (jeenbroekstra) [inline core/model/src/main/java/org/eclipse/rdf4j/model/impl/LinkedHashModel.java:205]: I assume this is here merely for testing/demonstration purposes, and not intended to be a fix for anything?

Comment (hmottestad) [reply core/model/src/main/java/org/eclipse/rdf4j/model/impl/LinkedHashModel.java:205]: Just increases the likelihood of the error occurring.

**Comment by hmottestad** [comment]
@jeenbroekstra I'm content with this fix now. I've fully encapsulated the models in the Changeset so that all interactions are synchronized. This meant a few little modifications other places in the code, but I'm very happy with how little needed to change. 

I've only tested the performance for adding data (not combinations of adding/removing/querying). So far performance looks unchanged, which is exactly what we want.

What remains?
 - [ ] remove yield
 - [ ] revert some of the test code, but keep the new exceptions printing (convert to logging though)

**Comment by hmottestad** [comment]
I think this is done now and ready for final review and merging. It only seems to affect SERIALIZABLE transactions on the NativeStore, so not really a priority to ship a fix.

**Comment by jeenbroekstra** [comment]
Looks as if the CI build timed out

**Comment by hmottestad** [comment]
Fixed now. Null pointer exception I had forgotten to handle.

**Pull request merged**